torchdynamo/guards.py

```diff
@@ -50,6 +60,98 @@
 )
 
 
+class NNModuleChangeTrackerUtil:
+    @staticmethod
+    def invalidate_module(module):
+        if module not in module_code_map:
+            # Module is out of scope / deleted (weak ref key ref dict)
+            return
+
+        for code in module_code_map[module]:
+            code.valid = False
+
+        del module_code_map[module]
+
+        if module not in module_to_guard_ids:
+            # Module is out of scope / deleted (weak ref key ref dict)
+            return
+
+        for object_id in module_to_guard_ids[module]:
+            module_associated_guarded_ids.remove(object_id)
+
+        del module_to_guard_ids[module]
+
+    @staticmethod
+    def setup(module, guarded_code):
+        module_id = id(module)
+        modulecls = module.__class__
+        if module_id not in module_to_guard_ids:
+            module_to_guard_ids[module] = set()
+
+        for buffer in module.buffers():
+            buffer_id = id(buffer)
+            module_associated_guarded_ids.add(buffer_id)
+            module_to_guard_ids[module].add(buffer_id)
+
+        for mod in module.modules():
+            mod_id = id(mod)
+            module_associated_guarded_ids.add(mod_id)
+            module_to_guard_ids[module].add(mod_id)
```

Comment:
What if you have two modules, both with the same submodule.  
- A has submodule C (in frame 1)
- B has submodule C (in frame 2)

Both A and B get guarded, then A gets invalidated.  

I think `module_associated_guarded_ids` should not be global, be local to a specific guard creation session.

Reply:
Yea, I agree. Is there an object whose lifecycle matches the guard creation session neatly? Perhaps GuardBuilder?